platforms/allwinner-d1/core/src/drivers/twi.rs

```diff
@@ -0,0 +1,809 @@
+            let res = if let Some(error) = guard.data.err.take() {
+                tracing::warn!(?error, ?dir, "TWI error");
+                Err(error)
+            } else {
+                match core::mem::replace(&mut guard.data.op, TwiOp::None) {
+                    TwiOp::Read { buf, .. } => {
+                        debug_assert_eq!(dir, OpKind::Read);
+                        Ok(buf)
+                    }
+                    TwiOp::Write { buf, .. } => {
+                        debug_assert_eq!(dir, OpKind::Write);
+                        Ok(buf)
+                    }
+                    _ => unreachable!(),
+                }
+            };
+            if rsp.send(res).is_err() {
+                tracing::trace!("I2C transaction handle dropped");
+                break;
+            }
+        }
+        // transaction ended!
+        tracing::debug!("I2C transaction ended");
+
+        let guard = self.isr.lock(self.twi);
+        guard.twi.twi_cntr.modify(|_r, w| {
+            w.m_stp().set_bit();
+            w
+        });
+    }
+}
+
+impl IsrData {
+    #[must_use]
+    fn lock<'a>(&'a self, twi: &'a twi::RegisterBlock) -> TwiDataGuard<'a> {
+        // disable TWI interrupts while holding the guard.
+        twi.twi_cntr.modify(|_r, w| w.int_en().low());
+        let data = unsafe { &mut *(self.data.get()) };
+        TwiDataGuard { data, twi }
+    }
+}
+
+impl Drop for TwiDataGuard<'_> {
+    fn drop(&mut self) {
+        self.twi.twi_cntr.modify(|_r, w| w.int_en().high());
+    }
+}
+
+impl TwiDataGuard<'_> {
+    async fn wait_for_irq(&mut self) {
+        let mut waiting = false;
+        futures::future::poll_fn(|cx| {
+            if waiting {
+                self.twi.twi_cntr.modify(|_r, w| w.int_en().low());
+                return Poll::Ready(());
+            }
+
+            self.data.waker = Some(cx.waker().clone());
+            waiting = true;
+            self.twi.twi_cntr.modify(|_r, w| {
+                // we have to set M_STA and A_ACK as part of the same write that
+                // sets the INT_EN bit, or else we will potentially do something
+                // weird if we do two separate TWI_CNTR writes. setting all of
+                // these now, atomically, avoids weird cases where we send a
+                // START for some random address, as far as i can tell.
+                w.m_sta().set_bit();
+                w.a_ack().set_bit();
+                w.int_en().high();
+                w.bus_en().respond();
+                w
+            });
+
+            Poll::Pending
+        })
+        .await;
+    }
+}
+
+impl Deref for TwiDataGuard<'_> {
+    type Target = TwiData;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.data
+    }
+}
+
+impl DerefMut for TwiDataGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.data
+    }
+}
+
+impl TwiData {
+    fn advance_isr<const NUM: u8>(&mut self, twi: &twi::RegisterBlock) {
```

Comment:
Just because this function is sorta big and I don't understand what NUM is used for, any reason to make this generic over NUM instead of taking it as an arg?

I dunno if const generics monomorphize bloat like other functions, but if we need the generic, we might want to make an inner function that isn't generic to keep code size down.

Reply:
i did the silly const generic thing because i thought there might be a performance benefit from not passing an extra arg that's only used for logging. but, code size probably has a bigger impact on performance than passing one integer in a register, so you're right, this is probably dumb!